Credentials look up

AlexSCorey opened the pull request:
##### SUMMARY
This PR addresses: https://github.com/ansible/awx/issues/4231.  @keithjgrant and I agreed that the `Lookup` component needs to be refactored (issue for that here: https://github.com/ansible/awx/issues/5065).  After this merges I plan to go and refactor the `Lookup` credential.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - UI

##### AWX VERSION
```
7.0.0
```


##### ADDITIONAL INFORMATION
![CredentialsLookup](https://user-images.githubusercontent.com/39280967/67327213-953cf900-f4e5-11e9-9d94-007fab950f37.gif)

AlexSCorey committed:
Adds CredentialLookUp to JT Form

AlexSCorey commented on awx/ui_next/src/components/Lookup/CredentialsLookup.jsx:
```diff
@@ -0,0 +1,165 @@
+};
+export { CredentialsLookup as _CredentialsLookup };
+
+export default withI18n()(withRouter(CredentialsLookup));
```
Oof.  Probably don't need to export `withRouter` here.  Will remove after getting feedback from others.

AlexSCorey commented on awx/ui_next/src/components/Lookup/CredentialsLookup.jsx:
```diff
@@ -0,0 +1,165 @@
+                  // require different field values.
+                  cred = {
+                    id: cred.id, kind: cred.kind, type: cred.namespace,
+                    value: cred.name, label: cred.name, isDisabled: false
```
Hopefully when I refactor `Lookup` some of the duplicated fields can be removed.

AlexSCorey commented on awx/ui_next/src/components/Lookup/Lookup.jsx:
```diff
@@ -110,30 +130,31 @@ class Lookup extends React.Component {
 
   toggleSelected(row) {
-    const { name, onLookupSave, multiple } = this.props;
+    const { name, onLookupSave, multiple, onToggleItem, selectCategoryOptions } = this.props;
```
I think the complexity in this function and be reduced once `Lookup` is refactored.

AlexSCorey commented on awx/ui_next/src/components/Lookup/Lookup.jsx:
```diff
@@ -177,11 +201,11 @@ class Lookup extends React.Component {
-      : lookupSelectedItems[0] || null;
-    onLookupSave(value, name);
-    this.handleModalToggle();
+    const value = multiple ? lookupSelectedItems : lookupSelectedItems[0] || null;
```
I plan to offload all of the state management back onto `Lookups` parents, but this work.

AlexSCorey committed:
Adds CredentialLookUp to JT Form

AlexSCorey committed:
Adds tests

keithjgrant commented on awx/ui_next/src/api/models/JobTemplates.js:
```diff
@@ -44,6 +44,19 @@ class JobTemplates extends InstanceGroupsMixin(NotificationsMixin(Base)) {
+
+  disassociateCredentials(id, credential) {
+    return this.http.post(`${this.baseUrl}${id}/credentials/`, {
+      id: credential,
```
Calling this variable `credentialId` instead of `credential` would be bit more clear

keithjgrant commented on awx/ui_next/src/components/Lookup/CredentialsLookup.jsx:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  componentDidMount() {
+    this.loadCredentials({ page: 1, page_size: 5, order_by: 'name' });
```
Why is this called here? This function doesn't do anything with the result plus `<Lookup>` below should be calling it directly (line 145)

AlexSCorey replied:
Good question.  Consider that function removed.

keithjgrant commented on awx/ui_next/src/components/Lookup/CredentialsLookup.jsx:
```diff
@@ -0,0 +1,173 @@
+    super(props);
+
+    this.state = {
+      credentials: props.credentials,
```
`credentials` missing in PropTypes below. Also, consider calling it `initialCredentials` (or `initialValue`?) so it's clear it's not a controlled component (though I would consider making it a controlled component and not setting credentials to state at all)

AlexSCorey replied:
I considered making this a controlled component, but chose not to because I didn't want to add to the state and required functions of `JobTemplateForm`, which is what would control this component.  Additionally, asking `JobTemplateForm` to control this component would also mean passing functions down to its grandchild, which in my opinion, isn't ideal if it can be avoided.

keithjgrant replied:
I don't think controlling this will add much to JT Form, since that value is managed by Formik

keithjgrant commented on awx/ui_next/src/components/Lookup/CredentialsLookup.jsx:
```diff
@@ -0,0 +1,173 @@
+    if (selectedCredentialType.value === 'Vault' || !credentialTypeOccupied) {
+      item.credentialType = selectedCredentialType;
+      this.setState({ credentials: [...stateToUpdate, item] });
+      onChange([...stateToUpdate, item]);
```
instead of creating two copies of the array, create it once to variable and pass that to `setState` and `onChange`

keithjgrant commented on awx/ui_next/src/components/Lookup/CredentialsLookup.test.jsx:
```diff
@@ -0,0 +1,103 @@
+    const chip = wrapper.find('PFChip').at(0);
+    expect(chip).toHaveLength(1);
+    chip.find('ChipButton').invoke('onClick')();
+    expect(wrapper.find('PFChip')).toHaveLength(1);
```
I don't understand this test... you make sure there's at least one chip, then remove it, and assert again that there's one chip? Should line 70 be asserting `toHaveLength(0)`?

AlexSCorey replied:
Good point. I was attempting to test that given 2 credentials on initial load, if I remove 1 there should still be 1.  I have improved that test now.  Thanks

keithjgrant commented on awx/ui_next/src/components/Lookup/Lookup.jsx:
```diff
@@ -76,21 +80,38 @@ class Lookup extends React.Component {
   componentDidMount() {
-    this.getData();
+    const { onLoadCredentialTypes } = this.props;
+    if (onLoadCredentialTypes) {
```
This prop isn't passed in from CredentialsLookup, so this `if` seems unnecessary

keithjgrant reviewed (changes requested):
Nice work on this! I haven't pulled it up in the UI yet since my containers are in a messy state right now, but I noted some things that jumped out to me in the code

AlexSCorey committed:
Addresses PR Issues

Improves credential ID variable in JT model.
Removes unused prop from Lookup ComponentDidMount.
Removed unused function call from Credentials ComponentDidMount.
Streamlines toggleCredential function and moves it to JobTemplateForm.  This was done because the
JobTemplateForm should handle the credential values passed to the CredentialsLookup.
Adds tests for JobTemplateForm to ensure toggleCredentialSelection function is putting proper values
in state.
Removed withRouter wrapper on CredentialsLookup export.
Improved CredentialsLookup test to ensure that onChange is called when user removes
a credential from the input.

AlexSCorey committed:
Fixes linting errors

keithjgrant commented on awx/ui_next/src/screens/Template/shared/JobTemplateForm.jsx:
```diff
@@ -104,6 +108,31 @@ class JobTemplateForm extends Component {
+        credential =>
+          credential.kind === 'vault' || credential.kind !== newCredential.kind
+      );
+      newCredentialsList = [...newCredentialsList, newCredential];
```
You can put all this logic in CredentialsLookup, then pass the full updated credentialsList to its `onChange` prop. Then in this component it's a simple `setFieldValue`. This will help keep the size of this form component down

keithjgrant commented on awx/ui_next/src/screens/Template/shared/JobTemplateForm.jsx:
```diff
@@ -319,6 +348,22 @@ class JobTemplateForm extends Component {
+            fieldId="template-credentials"
+            render={() => (
+              <CredentialsLookup
+                credentials={credentials}
```
Get `credentials` off the Formik field, then you don't need to track it in this component's state. See the LabelSelect on line 343 for an example

keithjgrant commented on awx/ui_next/src/components/Lookup/Lookup.jsx:
```diff
@@ -76,21 +80,33 @@ class Lookup extends React.Component {
 
   componentDidMount() {
-    this.getData();
+    Promise.all([this.getData()]);
```
`Promise.all()` isn't doing anything here—for two reasons: 1. nothing awaits it/there's no `.then()` here and 2. `getData()` only returns a single promise, so it doesn't need to be run through `Promise.all()` to await it.

keithjgrant commented on awx/ui_next/src/components/Lookup/Lookup.jsx:
```diff
@@ -76,21 +80,33 @@ class Lookup extends React.Component {
     }
   }
 
+  toggleDropdown() {
```
I don't see this being used anywhere. I think you can delete it and the `isDropdownOpen` state (AnsibleSelect manages its own dropdown open/closed state)

keithjgrant commented on awx/ui_next/src/components/Lookup/CredentialsLookup.test.jsx:
```diff
@@ -0,0 +1,90 @@
+    const button = chip.at(1).find('Button');
+    expect(chip).toHaveLength(2);
+    button.prop('onClick')();
+    expect(onChange).toBeCalledTimes(1);
```
You can use `.toBeCalledWith()` to assert the expected value is passed to `onChange` to make this a little more robust.

It might also be a good to add a check on the number of rendered chips (`expect(wrapper.find('Chip')).toHaveLength(1)`)

AlexSCorey committed:
Moves JT CredentialsList Manipulation Back to CredentialsLookup

Rename CredentialsLookup to MultiCredentialLookup
Removes unnecessary functions in Lookup.
Puts CredentialsList manipulation on CredsLookup and removes that work from JTForm.
Upates tests for CredentialsLookup and JTForm to reflect changes above.

keithjgrant commented on awx/ui_next/src/components/Lookup/MultiCredentialsLookup.jsx:
```diff
@@ -0,0 +1,150 @@
+}
+
+MultiCredentialsLookup.propTypes = {
+  tooltip: PropTypes.string,
```
Make sure you get the other props documented here: `onError`, `credentials`, `onChange`

keithjgrant commented on awx/ui_next/src/screens/Template/shared/JobTemplateForm.jsx:
```diff
@@ -319,6 +320,24 @@ class JobTemplateForm extends Component {
+              <MultiCredentialsLookup
+                credentials={field.value}
+                onChange={newCredentials =>
+                  setFieldValue('credentials', newCredentials)
```
👍🏻

keithjgrant reviewed (approved):
Looks good!

AlexSCorey committed:
Adds Proptypes
